pkg/declarative/v2/cleanup.go

```diff
@@ -0,0 +1,59 @@
+package v2
+
+import (
+	"context"
+
+	"github.com/kyma-project/module-manager/pkg/types"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/resource"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Cleanup interface {
+	Run(context.Context, []*resource.Info) (bool, error)
+}
+
+type ConcurrentCleanup struct {
+	clnt   client.Client
+	policy client.PropagationPolicy
+}
+
+func NewConcurrentCleanup(clnt client.Client) Cleanup {
+	return &ConcurrentCleanup{clnt: clnt, policy: client.PropagationPolicy(metav1.DeletePropagationBackground)}
+}
+
+func (c *ConcurrentCleanup) Run(ctx context.Context, infos []*resource.Info) (bool, error) {
+	// The Runtime Complexity of this Branch is N as only ServerSideApplier Patch is required
+	results := make(chan error, len(infos))
+	for i := range infos {
+		i := i
+		go c.cleanupResource(ctx, infos[i], results)
+	}
+
+	var errs []error
+	present := len(infos)
+	for i := 0; i < len(infos); i++ {
+		err := <-results
```

Comment:
is this some advanced feature of golang? we don't need to indicate the current index of the results and golang will take care of this?

Reply:
the results channel is bounded buffered so that means the size of the buffer is the same as the Len of infos. Thus you will always have exactly as many coroutines as infos

Reply:
ah, ok, in each iteration, the results are pulled out one by one, the order indexing is implied...